### PR TITLE
std: add test for child_process

### PR DESCRIPTION
**Pull request by matu3ba**
- Cli operations should be refactored, since the standard test runner
  has an expected argument structure. This would also ensure that the
  test cli is usable as tested library with checks for subprocess
  error or success instead of "hacky shell script interfaces".
- Default paths generation based on tmpDir would also be useful.
- Anonymous pipes on windows are generated from named pipes
- Async IO does not work on anonymous pipes
- Remove finished TODO

### CODE REVIEW

**Comment by matu3ba** [comment]
This could also be extended with capturing+comparing output of stdout and stderr ie with https://gist.github.com/matu3ba/0add09568327bad64e7181f9faadca40, but I think this should be good for now.

**Comment by matu3ba** [comment]
Windows fails with
```txt
75/1575 child_process.test "std-native-Debug-bare-multi build and call child_process"... FAIL (FileNotFound)
D:\a\1\s\lib\std\os\windows.zig:1601:32: 0x7ff7310fe35b in os.windows.CreateProcessW (test.obj)
            .FILE_NOT_FOUND => return error.FileNotFound,
```
MacOS x86_64 fails also with
```txt
378/2189 child_process.test "std-native-Debug--multi build and call child_process"... FAIL (FileNotFound)
FAIL (FileNotFound)
/Users/runner/work/1/s/lib/std/hash_map.zig:1470:13: 0x109c0bd3d in hash_map.HashMapUnmanaged([]const u8,void,hash_map.StringContext,80).cloneContext (test)
            try other.allocate(allocator, new_cap);
```
but succeeds on ARM64.
Same goes for sourcehut freebsd test runner
```txt
378/2189 child_process.test "std-native-Debug--multi build and call child_process"... FAIL (FileNotFound)
FAIL (FileNotFound)
```

More interestingly, the ci ziglang drone test runner fails with the wrong argument count?!
```txt
358/2195 child_process.test "std-native-Debug-bare-multi build and call child_process"... ./zig-cache/tmp/9PIXr2ojRhdmXBaF/child.zig:6:17: error: expected 2 argument(s), found 1
     _ = it.next() orelse unreachable; // skip binary name
                ^
/deps/local/lib/zig/std/process.zig:477:9: note: declared here
    pub fn next(self: *ArgIterator, allocator: Allocator) ?(NextError![:0]u8) {
        ^
expected 0, found 1
FAIL (TestExpectedEqual)
```

**Comment by matu3ba** [comment]
lavahut freebsd runner fails with
```txt
cwd_str: /usr/home/build/zig/build
tmpDir: wiqiIHyxnr16WBn8
child_zig: zig-cache/tmp/wiqiIHyxnr16WBn8/child.zig
emit_bin: -femit-bin=zig-cache/tmp/wiqiIHyxnr16WBn8/child
args compile: { zig, build-exe, zig-cache/tmp/wiqiIHyxnr16WBn8/child.zig, -femit-bin=zig-cache/tmp/wiqiIHyxnr16WBn8/child }
FAIL (FileNotFound)
```
since code assumes cwd is `/usr/home/build/zig`

ziglang ci drone runner fails with
```txt
cwd_str: /workspace
tmpDir: dH4qbEBkhRBdrE4s
child_zig: zig-cache/tmp/dH4qbEBkhRBdrE4s/child.zig
emit_bin: -femit-bin=zig-cache/tmp/dH4qbEBkhRBdrE4s/child
args compile: { zig, build-exe, zig-cache/tmp/dH4qbEBkhRBdrE4s/child.zig, -femit-bin=zig-cache/tmp/dH4qbEBkhRBdrE4s/child }
./zig-cache/tmp/dH4qbEBkhRBdrE4s/child.zig:6:17: error: expected 2 argument(s), found 1
     _ = it.next() orelse unreachable; // skip binary name
                ^
/deps/local/lib/zig/std/process.zig:477:9: note: declared here
    pub fn next(self: *ArgIterator, allocator: Allocator) ?(NextError![:0]u8) {
        ^
expected 0, found 1
FAIL (TestExpectedEqual)
```
which is interesting, because `std.debug.print("args child: {s}\n", .{args});` is for some reason not shown.

**Comment by matu3ba** [comment]
1. realpath is broken on freebsd due to `getFdPath` not workign `else => @compileError("querying for canonical path of a handle is unsupported on this host"),`
2. ci.ziglang.org/drone/pr still fails with
```txt
cwd_str: /workspace
tmpDir: m6JK1TJxFjzIpaO5
tmp_relpath: zig-cache/tmp/m6JK1TJxFjzIpaO5
tmp_abspath: /workspace/zig-cache/tmp/m6JK1TJxFjzIpaO5
child_zig: /workspace/zig-cache/tmp/m6JK1TJxFjzIpaO5/child.zig
emit_bin: -femit-bin=/workspace/zig-cache/tmp/m6JK1TJxFjzIpaO5/child
args compile: { zig, build-exe, /workspace/zig-cache/tmp/m6JK1TJxFjzIpaO5/child.zig, -femit-bin=/workspace/zig-cache/tmp/m6JK1TJxFjzIpaO5/child }
./zig-cache/tmp/m6JK1TJxFjzIpaO5/child.zig:6:17: error: expected 2 argument(s), found 1
     _ = it.next() orelse unreachable; // skip binary name
                ^
/deps/local/lib/zig/std/process.zig:477:9: note: declared here
    pub fn next(self: *ArgIterator, allocator: Allocator) ?(NextError![:0]u8) {
        ^
expected 0, found 1
FAIL (TestExpectedEqual)
```
However, there is no code, ie  `rg -F 'pub fn next(self: *ArgIterator, allocator: Allocator) ?(NextError![:0]u8) {'` shows no results.
This is very very weird and looks like libstd may be wrongly installed or cached and is using the version before the argument iterator change (which curiously I cant find with `git grep -w 'pub fn next(self: *ArgIterator, allocator: Allocator) ?(NextError![:0]u8) {'`
3. Windows: wrong delimitor and probably Windows being annoying and requires `.exe` for executables

**Comment by matu3ba** [comment]
works. next: replace `realpath` either with relative path or concatenated one.

**Comment by matu3ba** [comment]
ready for review/to be merged.

**Comment by matu3ba** [comment]
now ready for review/merge. fixed a typo.

File: lib/std/child_process.zig [hunk 1]
```diff
@@ -1201,3 +1202,69 @@ test "createNullDelimitedEnvMap" {
...
+    const cache = "zig-cache";
+    const tmpdir = "tmp";
+    const child_name = "child"; // no need for suffixes (.exe, .wasm) due to '-femit-bin'
+    const suffix_zig = ".zig";
+    const child_path = try fs.path.join(testing.allocator, &[_][]const u8{ cwd_str, std.fs.path.sep_str, cache, tmpdir, &tmp.sub_path, child_name });
+    defer testing.allocator.free(child_path);
+
+    const child_zig = try mem.concat(testing.allocator, u8, &[_][]const u8{ child_path, suffix_zig });
+    defer testing.allocator.free(child_zig);
+    //std.debug.print("child_zig: {s}\n", .{child_zig});
```
Comment (Vexu) [inline lib/std/child_process.zig:1245]: Stray debug print?

File: lib/std/child_process.zig [hunk 2]
```diff
@@ -1201,3 +1202,69 @@ test "createNullDelimitedEnvMap" {
...
+    {
+        // 'zigexec build-exe path/to/child.zig -femit-bin=path/to/child' expect success
+        try tmp.dir.writeFile("child.zig", childstr);
+        const args = [_][]const u8{ zigexec, "build-exe", child_zig, emit_bin };
+        var procCompileChild = try ChildProcess.init(&args, testing.allocator);
+        defer procCompileChild.deinit();
+        try procCompileChild.spawn();
+        const ret_val = try procCompileChild.wait();
+        try testing.expectEqualSlices(u8, exited, std.meta.tagName(ret_val));
+        try testing.expectEqual(@as(u8, 0), ret_val.Exited);
```
Comment (Vexu) [inline lib/std/child_process.zig:1258]:
```suggestion
        try testing.expectEqual(ret_val, .{ .Exited = 0 });
```

File: lib/std/child_process.zig [hunk 3]
```diff
@@ -1201,3 +1202,69 @@ test "createNullDelimitedEnvMap" {
...
+    }
+    {
+        // spawn compiled file as child_process with argument 'hello world' + expect success
+        const args = [_][]const u8{ child_path, "hello world" };
+        var child_proc = try ChildProcess.init(&args, testing.allocator);
+        defer child_proc.deinit();
+        try child_proc.spawn();
+        const ret_val = try child_proc.wait();
+        try testing.expectEqualSlices(u8, exited, std.meta.tagName(ret_val));
+        try testing.expectEqual(@as(u8, 0), ret_val.Exited);
```
Comment (Vexu) [inline lib/std/child_process.zig:1268]: Ditto.

**Review by Vexu — APPROVED** [review]

File: lib/std/child_process.zig (hunk 1 above)
Comment (matu3ba) [reply lib/std/child_process.zig:1245]: ups.

File: lib/std/child_process.zig (hunk 2 above)
Comment (matu3ba) [reply lib/std/child_process.zig]: Fixed.

File: lib/std/child_process.zig (hunk 3 above)
Comment (matu3ba) [reply lib/std/child_process.zig]: fixed

**Pull request merged**